src/sshfs_mount/sshfs_mounts.cpp

```diff
@@ -162,15 +167,14 @@ void mp::SSHFSMounts::stop_all_mounts_for_instance(const std::string& instance)
     auto mounts_it = mount_processes.find(instance);
     if (mounts_it == mount_processes.end() || mounts_it->second.empty())
     {
-        mpl::log(mpl::Level::debug, category, fmt::format("No mounts to stop for instance \"{}\"", instance));
+        mpl::log(mpl::Level::info, category, fmt::format("No mounts to stop for instance \"{}\"", instance));
     }
     else
     {
-        for (auto& sshfs_mount : mounts_it->second)
+        for (auto it = mounts_it->second.cbegin(), next = it; it != mounts_it->second.cend(); it = next)
         {
-            mpl::log(mpl::Level::debug, category,
-                     fmt::format("Stopping mount '{}' in instance \"{}\"", sshfs_mount.first, instance));
-            sshfs_mount.second->terminate();
+            next++;
+            stop_mount(instance, it->first);
         }
```

Comment:
i don't really get this for-loop. what is the point of `next`? can't this be simplified like this?
```cpp
for (auto it = mounts_it->second.cbegin(), next = it; it != mounts_it->second.cend(); it = next)
{
    next++;
    stop_mount(instance, it->first);
}
```
:arrow_down: 
```cpp
for (auto it = mounts_it->second.cbegin(), next = it; it != mounts_it->second.cend(); it = next)
{
    stop_mount(instance, it->first);
    next++;
}
```
:arrow_down: 
```cpp
for (auto it = mounts_it->second.cbegin(), next = it; it != mounts_it->second.cend(); it = ++next)
    stop_mount(instance, it->first);
```
:arrow_down: 
```cpp
for (auto it = mounts_it->second.cbegin(); it != mounts_it->second.cend(); it++)
    stop_mount(instance, it->first);
```
:arrow_down: 
```cpp
for (const auto& it : mounts_it->second)
    stop_mount(instance, it->first);
```
:arrow_down: 
```cpp
for (const auto& [path, _] : mounts_it->second)
    stop_mount(instance, path);
```

Reply:
The purpose behind `next` is to keep track of the position of the next item in the `mount_processes` map. When `stop_mount` is called, the connection listener will delete the item from the map [here](https://github.com/canonical/multipass/blob/645f9df7462cf10f63d4568cd66804817c0ba31b/src/sshfs_mount/sshfs_mounts.cpp#L101) thus losing a pointer to the next item. `next` will keep track of the next item eliminating a potential race.

Reply:
Ah, it's about iterator invalidation in maps! Now I understand, thanks for explaining!
Can you please leave a comment above the loop explaining the reason behind it? I'm sure that will be helpful in the future.
Also since this is about keeping the iterator one step ahead of the deletes, how about
```cpp
for (auto it = mounts_it->second.cbegin(); it != mounts_it->second.cend();)
    stop_mount(instance, it++->path);
```

Reply:
Made your suggested changes to the loop and also added a comment.